File: scripts/scenario_editor/openScenario/maneuver_entities.py
def accelerate_all_simulation_cars(root, storyboard, number_of_sim_cars, speed, start_after_x_sec, stop_after_x_sec):
    story = root.createElement('Story')
    story.setAttribute('name', 'MyStory')
    act = root.createElement('Act')
    act.setAttribute('name', 'Behavior')
    maneuver_group = root.createElement('ManeuverGroup')
    maneuver_group.setAttribute('maximumExecutionCount', '1')
    maneuver_group.setAttribute('name', 'ManeuverSequence')
    actors = root.createElement('Actors')
    actors.setAttribute('selectTriggeringEntities', 'false')
    for n in range(0, number_of_sim_cars): # which vehicles are affected (all sim. cars)
        entity_ref = root.createElement('EntityRef')
        entity_ref.setAttribute('entityRef', 'adversary'+str(n))
        actors.appendChild(entity_ref)
    maneuver = root.createElement('Maneuver')
    maneuver.setAttribute('name', 'AccelerateManeuver')

    # What this maneuver is doing --> Accelerate cars
    event = root.createElement('Event')
    event.setAttribute('name', 'AdversaryAccelerates')
    event.setAttribute('priority', 'overwrite')
    action = root.createElement('Action')
    action.setAttribute('name', 'AdversaryAccelerates')
    private_action = root.createElement('PrivateAction')
    longitudinal_action = root.createElement('LongitudinalAction')
    speed_action = root.createElement('SpeedAction')
    speed_action_dynamics = root.createElement('SpeedActionDynamics')
    speed_action_dynamics.setAttribute('dynamicsShape', 'step')
    speed_action_dynamics.setAttribute('value', '50')
    speed_action_dynamics.setAttribute('dynamicsDimension', 'distance')
    speed_action_target = root.createElement('SpeedActionTarget')
    absolute_target_speed = root.createElement('AbsoluteTargetSpeed')
    absolute_target_speed.setAttribute('value', str(speed))
    
    # Start condition
    start_trigger = root.createElement('StartTrigger')
    accelerate_condition(root, start_trigger, "StartTime", start_after_x_sec)

    # Stop condition
    stop_trigger = root.createElement('StopTrigger')
    accelerate_condition(root, stop_trigger, "StopTime", stop_after_x_sec)
    
    # Hierarchy
    storyboard.appendChild(story)
    story.appendChild(act)
    act.appendChild(maneuver_group)
    maneuver_group.appendChild(actors)
    # for n in range(0, number_of_sim_cars):
    #     actors.appendChild(entity_ref[n])
    maneuver_group.appendChild(maneuver)
    maneuver.appendChild(event)
    event.appendChild(action)
    action.appendChild(private_action)
    private_action.appendChild(longitudinal_action)
    longitudinal_action.appendChild(speed_action)
    speed_action.appendChild(speed_action_dynamics)
    speed_action.appendChild(speed_action_target)
    speed_action_target.appendChild(absolute_target_speed)
    event.appendChild(start_trigger)
    act.appendChild(root.createElement('StartTrigger'))
    act.appendChild(stop_trigger)
    storyboard.appendChild(root.createElement('StopTrigger'))

def accelerate_condition(root, trigger, name, time_in_sec):
    condition_group = root.createElement('ConditionGroup')
    condition = root.createElement('Condition')
    condition.setAttribute('name', str(name))
    condition.setAttribute('delay', '0')
    condition.setAttribute('conditionEdge', 'rising')
    by_value_condition = root.createElement('ByValueCondition')
    simulation_time_condition = root.createElement('SimulationTimeCondition')
    simulation_time_condition.setAttribute('value', str(time_in_sec))
    simulation_time_condition.setAttribute('rule', 'greaterThan')

    trigger.appendChild(condition_group)
    condition_group.appendChild(condition)
    condition.appendChild(by_value_condition)
    by_value_condition.appendChild(simulation_time_condition)

    

File: C#/OpenScenario/OpenScenario/ManeuverEntities.cs
using System.Xml;
class ManeuverEntities
{

    public XmlDocument root;

    public ManeuverEntities(XmlDocument root)
    {
        this.root = root;
    }

    public void AccelerateAllSimulationCars(XmlNode storyboard, int number_of_sim_cars, double speed, int start_after_x_sec, int stop_after_x_sec)
    {
        XmlNode story = root.CreateElement("Story");
        SetAttribute("name", "MyStory", story);
        XmlNode act = root.CreateElement("Act");
        SetAttribute("name", "Behavior", act);
        XmlNode maneuver_group = root.CreateElement("ManeuverGroup");
        SetAttribute("maximumExecutionCount", "1", maneuver_group);
        SetAttribute("name", "ManeuverSequence", maneuver_group);
        XmlNode actors = root.CreateElement("Actors");
        SetAttribute("selectTriggeringEntities", "false", actors);
        for (int n = 0; n < number_of_sim_cars; n++) // which vehicles are affected (all sim. cars)
        {
            XmlNode entity_ref = root.CreateElement("EntityRef");
            SetAttribute("entityRef", "adversary" + n.ToString(), entity_ref);
            actors.AppendChild(entity_ref);
        }
        XmlNode maneuver = root.CreateElement("Maneuver");
        SetAttribute("name", "AccelerateManeuver", maneuver);

        // What this maneuver is doing --> Accelerate cars
        XmlNode _event = root.CreateElement("Event");
        SetAttribute("name", "AdversaryAccelerates", _event);
        SetAttribute("priority", "overwrite", _event);
        XmlNode action = root.CreateElement("Action");
        SetAttribute("name", "AdversaryAccelerates", action);
        XmlNode private_action = root.CreateElement("PrivateAction");
        XmlNode longitudinal_action = root.CreateElement("LongitudinalAction");
        XmlNode speed_action = root.CreateElement("SpeedAction");
        XmlNode speed_action_dynamics = root.CreateElement("SpeedActionDynamics");
        SetAttribute("dynamicsShape", "step", speed_action_dynamics);
        SetAttribute("value", "50", speed_action_dynamics);
        SetAttribute("dynamicsDimension", "distance", speed_action_dynamics);
        XmlNode speed_action_target = root.CreateElement("SpeedActionTarget");
        XmlNode absolute_target_speed = root.CreateElement("AbsoluteTargetSpeed");
        SetAttribute("value", speed.ToString(), absolute_target_speed);

        // Start condition
        XmlNode start_trigger = root.CreateElement("StartTrigger");
        AccelerateCondition(start_trigger, "StartTime", start_after_x_sec);

        // Stop condition
        XmlNode stop_trigger = root.CreateElement("StopTrigger");
        AccelerateCondition(stop_trigger, "StopTime", stop_after_x_sec);

        // Hierarchy
        storyboard.AppendChild(story);
        story.AppendChild(act);
        act.AppendChild(maneuver_group);
        maneuver_group.AppendChild(actors);
        maneuver_group.AppendChild(maneuver);
        maneuver.AppendChild(_event);
        _event.AppendChild(action);
        action.AppendChild(private_action);
        private_action.AppendChild(longitudinal_action);
        longitudinal_action.AppendChild(speed_action);
        speed_action.AppendChild(speed_action_dynamics);
        speed_action.AppendChild(speed_action_target);
        speed_action_target.AppendChild(absolute_target_speed);
        _event.AppendChild(start_trigger);
        act.AppendChild(root.CreateElement("StartTrigger"));
        act.AppendChild(stop_trigger);
        storyboard.AppendChild(root.CreateElement("StopTrigger"));
    }
        
    private void AccelerateCondition(XmlNode trigger, String name, int time_in_sec)
    {
        XmlNode condition_group = root.CreateElement("ConditionGroup");
        XmlNode condition = root.CreateElement("Condition");
        SetAttribute("name", name.ToString(), condition);
        SetAttribute("delay", "0", condition);
        SetAttribute("conditionEdge", "rising", condition);
        XmlNode by_value_condition = root.CreateElement("ByValueCondition");
        XmlNode simulation_time_condition = root.CreateElement("SimulationTimeCondition");
        SetAttribute("value", time_in_sec.ToString(), simulation_time_condition);
        SetAttribute("rule", "greaterThan", simulation_time_condition);

        trigger.AppendChild(condition_group);
        condition_group.AppendChild(condition);
        condition.AppendChild(by_value_condition);
        by_value_condition.AppendChild(simulation_time_condition);
    }
    private void SetAttribute(String name, String value, XmlNode element)
    {
        XmlAttribute attribute = root.CreateAttribute(name);
        attribute.Value = value;
        element.Attributes.Append(attribute);
    }

}







